ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -12193,6 +12188,22 @@ samplePredicate, true, new SampleDesc(ts.getNumerator(),
     return output;
   }
 
+  public List<VirtualColumn> getVirtualColumns(Table tab) {
+    List<VirtualColumn> virtualColumns = new ArrayList<>();
+    if (tab.isNative()) {
+      virtualColumns.addAll(VirtualColumn.getRegistry(conf));
+    }
+    if (tab.isNonNative() && AcidUtils.isNonNativeAcidTable(tab)) {
+      virtualColumns.addAll(tab.getStorageHandler().acidVirtualColumns());
+    }
+    if (tab.isNonNative() && tab.getStorageHandler().areSnapshotsSupported() &&
+        isBlank(tab.getMetaTable())) {
+      virtualColumns.add(VirtualColumn.SNAPSHOT_ID);
+    }
+    
+    return virtualColumns;
+  }
+
```

Comment:
nit: I like the refactoring although I feel that the method may have better encapsulation if moved inside `Table`.